File: gui/install_command_allowlist.go
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"log"
	"strings"
	"sync"
)

// Shared allowlist for AI-assistant install slash commands.
// Source of truth (also imported by the frontend):
//
//	frontend/src/components/ai/installCommandAllowlist.json
//
//go:embed frontend/src/components/ai/installCommandAllowlist.json
var installCommandAllowlistJSON []byte

// Nested action groups may carry aliases (e.g. market → marketplace).
type installNestedSpec struct {
	Aliases []string `json:"aliases"`
	Actions []string `json:"actions"`
}

type installCommandSpec struct {
	Aliases []string                     `json:"aliases"`
	Actions []string                     `json:"actions"`
	Nested  map[string]installNestedSpec `json:"nested"`
}

type installCommandAllowlistFile struct {
	Version        int                           `json:"version"`
	BinaryPrefixes []string                      `json:"binary_prefixes"`
	MetaActions    []string                      `json:"meta_actions"`
	Commands       map[string]installCommandSpec `json:"commands"`
}

type installCommandAllowlist struct {
	// canonical command name → action set
	actions map[string]map[string]struct{}
	// nested[cmd][action] → sub-action set (e.g. plugin/marketplace/add)
	nested map[string]map[string]map[string]struct{}
	// alias or canonical → canonical
	normalize map[string]string
	// meta actions always allowed (help flags)
	meta map[string]struct{}
	// CLI binary first tokens (without .exe)
	binaries map[string]struct{}
}

var (
	installAllowlistOnce sync.Once
	installAllowlistVal  *installCommandAllowlist
	installAllowlistErr  error
)

func loadInstallCommandAllowlist() (*installCommandAllowlist, error) {
	installAllowlistOnce.Do(func() {
		var raw installCommandAllowlistFile
		if err := json.Unmarshal(installCommandAllowlistJSON, &raw); err != nil {
			installAllowlistErr = fmt.Errorf("parse install command allowlist: %w", err)
			return
		}
		if len(raw.Commands) == 0 {
			installAllowlistErr = fmt.Errorf("install command allowlist has no commands")
			return
		}
		al := &installCommandAllowlist{
			actions:   make(map[string]map[string]struct{}, len(raw.Commands)),
			nested:    make(map[string]map[string]map[string]struct{}),
			normalize: make(map[string]string),
			meta:      installAllowlistStringSet(raw.MetaActions),
			binaries:  make(map[string]struct{}, len(raw.BinaryPrefixes)),
		}
		for _, b := range raw.BinaryPrefixes {
			b = strings.ToLower(strings.TrimSpace(b))
			b = strings.TrimSuffix(b, ".exe")
			if b != "" {
				al.binaries[b] = struct{}{}
			}
		}
		for name, spec := range raw.Commands {
			canonical := strings.ToLower(strings.TrimSpace(name))
			if canonical == "" {
				continue
			}
			al.normalize[canonical] = canonical
			for _, alias := range spec.Aliases {
				a := strings.ToLower(strings.TrimSpace(alias))
				if a != "" {
					al.normalize[a] = canonical
				}
			}
			acts := installAllowlistStringSet(spec.Actions)
			for m := range al.meta {
				acts[m] = struct{}{}
			}

			if len(spec.Nested) > 0 {
				nestedActs := make(map[string]map[string]struct{}, len(spec.Nested)*2)
				for parent, nestedSpec := range spec.Nested {
					p := strings.ToLower(strings.TrimSpace(parent))
					if p == "" {
						continue
					}
					set := installAllowlistStringSet(nestedSpec.Actions)
					for m := range al.meta {
						set[m] = struct{}{}
					}
					// Nested parents (and aliases) are also top-level actions,
					// so JSON need not list "marketplace" twice.
					acts[p] = struct{}{}
					nestedActs[p] = set
					for _, alias := range nestedSpec.Aliases {
						a := strings.ToLower(strings.TrimSpace(alias))
						if a == "" {
							continue
						}
						acts[a] = struct{}{}
						nestedActs[a] = set
					}
				}
				al.nested[canonical] = nestedActs
			}
			al.actions[canonical] = acts
		}
		installAllowlistVal = al
	})
	return installAllowlistVal, installAllowlistErr
}

func installAllowlistStringSet(values []string) map[string]struct{} {
	out := make(map[string]struct{}, len(values))
	for _, v := range values {
		v = strings.ToLower(strings.TrimSpace(v))
		if v != "" {
			out[v] = struct{}{}
		}
	}
	return out
}

func emptyInstallAllowlist() *installCommandAllowlist {
	return &installCommandAllowlist{
		actions:   map[string]map[string]struct{}{},
		nested:    map[string]map[string]map[string]struct{}{},
		normalize: map[string]string{},
		meta:      map[string]struct{}{},
		binaries:  map[string]struct{}{},
	}
}

var installAllowlistFailLogged sync.Once

func mustInstallAllowlist() *installCommandAllowlist {
	al, err := loadInstallCommandAllowlist()
	if err != nil || al == nil {
		installAllowlistFailLogged.Do(func() {
			// Avoid spamming; one warning is enough for operators.
			log.Printf("[install-allowlist] failed to load shared allowlist: %v", err)
		})
		// Fail closed: empty allowlist rejects everything.
		return emptyInstallAllowlist()
	}
	return al
}

// isKnownInstallCommand reports whether name (or alias) is a known install command.
func isKnownInstallCommand(name string) bool {
	al := mustInstallAllowlist()
	_, ok := al.normalize[strings.ToLower(strings.TrimSpace(name))]
	return ok
}

// isInstallMetaAction reports whether action is a global help/meta flag.
func isInstallMetaAction(action string) bool {
	al := mustInstallAllowlist()
	_, ok := al.meta[strings.ToLower(strings.TrimSpace(action))]
	return ok
}

// normalizeInstallCmd maps aliases (skills/plugins) to canonical command names.
func normalizeInstallCmd(head string) string {
	al := mustInstallAllowlist()
	key := strings.ToLower(strings.TrimSpace(head))
	if canon, ok := al.normalize[key]; ok {
		return canon
	}
	return key
}

// isInstallCLIBinaryPrefix reports whether token is a known CLI binary name
// (maclaw-tui, maclaw.exe, …) used when pasting full CLI lines into chat.
// Accepts bare names and path basenames (e.g. C:\bin\maclaw-tui.exe).
func isInstallCLIBinaryPrefix(token string) bool {
	al := mustInstallAllowlist()
	t := installCLIBinaryBaseName(token)
	_, ok := al.binaries[t]
	return ok
}

func installCLIBinaryBaseName(token string) string {
	t := strings.ToLower(strings.TrimSpace(token))
	// Strip surrounding quotes from paste.
	if len(t) >= 2 {
		if (t[0] == '"' && t[len(t)-1] == '"') || (t[0] == '\'' && t[len(t)-1] == '\'') {
			t = t[1 : len(t)-1]
		}
	}
	t = strings.TrimSuffix(t, ".exe")
	if i := strings.LastIndexAny(t, `/\`); i >= 0 {
		t = t[i+1:]
	}
	t = strings.TrimSuffix(t, ".exe")
	return strings.TrimSpace(t)
}

// isInstallNestedParent reports whether action is a nested parent for cmd
// (e.g. plugin + marketplace / market).
func isInstallNestedParent(cmd, action string) bool {
	al := mustInstallAllowlist()
	cmd = normalizeInstallCmd(cmd)
	action = strings.ToLower(strings.TrimSpace(action))
	nested, ok := al.nested[cmd]
	if !ok {
		return false
	}
	_, ok = nested[action]
	return ok
}

// installActionAllowed reports whether cmd+args is an allowed install slash action.
// cmd may be an alias; it is normalized internally.
func installActionAllowed(cmd string, args []string) bool {
	if len(args) == 0 {
		return false
	}
	al := mustInstallAllowlist()
	cmd = normalizeInstallCmd(cmd)
	action := strings.ToLower(strings.TrimSpace(args[0]))

	acts, ok := al.actions[cmd]
	if !ok {
		return false
	}
	// Meta help is allowed only for known install commands.
	if _, ok := al.meta[action]; ok {
		return true
	}
	if _, ok := acts[action]; !ok {
		return false
	}

	// Nested subcommands (e.g. plugin marketplace add).
	if nested, ok := al.nested[cmd]; ok {
		if subSet, ok := nested[action]; ok {
			// Bare `/plugin marketplace` is allowed (shows usage).
			if len(args) == 1 {
				return true
			}
			sub := strings.ToLower(strings.TrimSpace(args[1]))
			_, ok := subSet[sub]
			return ok
		}
	}
	return true
}

File: guiapp/frontend/src/components/ai/installCommandAllowlist.ts
/**
 * Shared install-command allowlist helpers for the AI assistant panel.
 *
 * Source of truth (also embedded by the Go backend via go:embed):
 *   ./installCommandAllowlist.json
 */
import allowlist from "./installCommandAllowlist.json";

export type InstallCommandAllowlist = typeof allowlist;

export const installCommandAllowlist: InstallCommandAllowlist = allowlist;

type NestedSpec = {
    aliases?: string[];
    actions?: string[];
};

function lowerSet(values: string[] | undefined): Set<string> {
    const set = new Set<string>();
    for (const v of values ?? []) {
        const key = v.trim().toLowerCase();
        if (key) set.add(key);
    }
    return set;
}

const META_ACTIONS = lowerSet(allowlist.meta_actions);

const COMMAND_ALIASES = new Map<string, string>();
const COMMAND_ACTIONS = new Map<string, Set<string>>();
/** nested[cmd][parentAction] → allowed sub-actions */
const COMMAND_NESTED = new Map<string, Map<string, Set<string>>>();

for (const [name, spec] of Object.entries(allowlist.commands)) {
    const canonical = name.trim().toLowerCase();
    COMMAND_ALIASES.set(canonical, canonical);
    for (const alias of spec.aliases ?? []) {
        const a = alias.trim().toLowerCase();
        if (a) COMMAND_ALIASES.set(a, canonical);
    }
    const acts = lowerSet(spec.actions);
    for (const m of META_ACTIONS) acts.add(m);

    const nestedRaw = "nested" in spec ? (spec.nested as Record<string, NestedSpec>) : undefined;
    if (nestedRaw) {
        const nested = new Map<string, Set<string>>();
        for (const [parent, value] of Object.entries(nestedRaw)) {
            const aliases = value?.aliases ?? [];
            const actions = value?.actions ?? [];
            const set = lowerSet(actions);
            for (const m of META_ACTIONS) set.add(m);
            const p = parent.trim().toLowerCase();
            // Nested parents (and aliases) are also top-level actions.
            acts.add(p);
            nested.set(p, set);
            for (const alias of aliases) {
                const a = alias.trim().toLowerCase();
                if (!a) continue;
                acts.add(a);
                nested.set(a, set);
            }
        }
        COMMAND_NESTED.set(canonical, nested);
    }

    COMMAND_ACTIONS.set(canonical, acts);
}

const BINARY_PREFIXES = lowerSet(
    (allowlist.binary_prefixes ?? []).map((p) => p.replace(/\.exe$/i, "")),
);

/** Regex matching a leading CLI binary token + space (e.g. `maclaw-tui `). */
export const INSTALL_BIN_PREFIX: RegExp = (() => {
    const parts = [...BINARY_PREFIXES]
        .map((p) => p.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"))
        .sort((a, b) => b.length - a.length);
    if (parts.length === 0) {
        return /^(?!)/; // never matches
    }
    return new RegExp(`^(?:${parts.join("|")})(?:\\.exe)?\\s+`, "i");
})();

export function normalizeInstallCommand(name: string): string {
    const key = name.trim().toLowerCase();
    return COMMAND_ALIASES.get(key) ?? key;
}

export function isKnownInstallCommand(name: string): boolean {
    return COMMAND_ALIASES.has(name.trim().toLowerCase());
}

export function isInstallMetaAction(action: string): boolean {
    return META_ACTIONS.has(action.trim().toLowerCase());
}

/** Whether action is a nested parent for cmd (e.g. plugin + marketplace). */
export function isInstallNestedParent(cmd: string, action: string): boolean {
    const nested = COMMAND_NESTED.get(normalizeInstallCommand(cmd));
    return !!nested?.has(action.trim().toLowerCase());
}

/**
 * Whether cmd+args is an allowed install slash action.
 * Mirrors Go installActionAllowed() from the same JSON allowlist.
 */
export function isInstallActionAllowed(cmd: string, args: string[]): boolean {
    if (!args.length) return false;
    const canonical = normalizeInstallCommand(cmd);
    const action = args[0].trim().toLowerCase();

    const acts = COMMAND_ACTIONS.get(canonical);
    if (!acts) return false;
    // Meta help is allowed only for known install commands.
    if (META_ACTIONS.has(action)) return true;
    if (!acts.has(action)) return false;

    const nested = COMMAND_NESTED.get(canonical);
    if (nested?.has(action)) {
        if (args.length === 1) return true; // bare parent shows usage
        const sub = args[1].trim().toLowerCase();
        return nested.get(action)!.has(sub);
    }
    return true;
}

/** Basename of a CLI token (strips path + .exe + surrounding quotes). */
export function installCLIBinaryBaseName(token: string): string {
    let t = token.trim().toLowerCase();
    if (
        (t.startsWith('"') && t.endsWith('"')) ||
        (t.startsWith("'") && t.endsWith("'"))
    ) {
        t = t.slice(1, -1);
    }
    t = t.replace(/\.exe$/i, "");
    const slash = Math.max(t.lastIndexOf("/"), t.lastIndexOf("\\"));
    if (slash >= 0) t = t.slice(slash + 1);
    return t.replace(/\.exe$/i, "").trim();
}

export function isInstallCLIBinaryPrefix(token: string): boolean {
    return BINARY_PREFIXES.has(installCLIBinaryBaseName(token));
}
